### PR TITLE
More response types

### PR DESCRIPTION
**Pull request by cap10morgan**
[Pivotal card](https://www.pivotaltracker.com/story/show/144867263)

This is to better support "this user already exists" responses for TV Classic user imports. I started with `409 Conflict` responses (so that's why that is added in here), but then realized in this context it shouldn't be an error. We're sending lots of user create requests fully anticipating that some will already exist and should be left alone. So that's just a different success scenario. That's how I ended up with a `200 OK` for the `:duplicate` status.

This has been tested in krakenstein with TV Classic user imports.

### CODE REVIEW

File: src/bifrost/core.clj
Comment (gws) [inline src/bifrost/core.clj]: Unused `log` reference?

**Review by gws — APPROVED** [review]
:+1: if that unused log reference is out.

My feeling is if the operation for adding a duplicate record is idempotent, then the API call shouldn't distinguish between the cases at all (e.g. just return `{:status :ok}`). However, I'm ultimately neutral on that point in terms of this PR because there might be a good reason that's not evident here, and because if that were to change in the future, the same HTTP status code would ultimately be returned.

File: src/bifrost/core.clj
Comment (cap10morgan) [reply src/bifrost/core.clj]: Oops, yeah, I'll get rid of that.

**Comment by cap10morgan** [comment]
Hmm... yeah, that could work too. This is for a POST to `/users` to create new users, and when you have `:reject-if-exists true` in your map, it will behave idempotently. I also debated moving the "reject dupes" flow over to a PUT request (which are typically idempotent vs. POST typically not) but I think that would have required changes to turbovote-web too to make sure it started using PATCH for all user updates (which would be more correct anyway).

Ultimately I'm distinguishing based on the HTTP response code either way for the import UX (201 for new user created, 200 for dupes). I settled on that because I didn't see a good alternative 2xx response code for "already exists" and I suspect that's by design. If so that argues for your approach too.

Do you think it would make it unnecessarily hard to tell what the result was for RabbitMQ interactions if the response was `:status :created` vs. `:status :ok` (on dupes)? That feels like the kind of thing that will cause someone to waste half a day down the road (even if we document it) thinking that the `:ok` meant the user they were trying to create got created when it actually wasn't. I think that's my only remaining hesitation.

**Comment by gws** [comment]
Ah, I see. I don't know that my comments are enough to hold up the PR, but here are my thoughts.

I think `:reject-if-exists true` should result in a... rejection, and a 409 response code over HTTP. A client would just anticipate this and behave accordingly.

If `:reject-if-exists false` is present, I suppose `:status :duplicate` is fine, as well as a 200 OK response. It's distinct from `:created` and the client seems to be accepting that nothing will change.

I do think it'd make it hard to tell what happened if the client is expecting to see their data reflected after a POST to `/users` and they do not see it if they set `:reject-if-exists true`. If they set `:reject-if-exists false` then they seem to be accepting the consequences, so in that case a 200 OK response seems fine.

**Comment by gws** [comment]
I think my preferred path would be to do away with the `:reject-if-exists` flag entirely, always reject "duplicate" POSTs with a body that contains enough information for the clients to resolve the conflict, and clients can do an explicit PUT using that information if they want to overwrite it. That is out of scope of the PR and maybe not desired or possible here, so the `:status :duplicate` situation above may be fine here.

**Comment by cap10morgan** [comment]
Yeah, I agree that `:reject-if-exists true` implies an error if it exists. The main problem with doing it that way now is that classic migrations would flood us with innocuous "errors" which I want to avoid at all costs. If I had this to re-design all over again... ;)

`:reject-if-exists false` (or really, missing altogether) means "ignore existing users and create this one even if it's a dupe" because that's the current behavior we needed to preserve for the turbovote-web usage of this API. So that will never return a `:status :duplicate`.

The 2 options currently are:

1. `:reject-if-exists true` - enable dupe checking
1. `:reject-if-exists false` or not present - disable dupe checking

At this point I'm well into pragmatic "get this thing out and usable" territory given how soon we're trying to start migrating users. I think bigger changes to how we're approaching this might have to wait until future refactorings.

But it looks like I have some reconciliation work ahead of me with the recent landing of the create-session feature of user-works anyway. So I'll see what comes of that before merging this.

Thanks for the review and thoughtful feedback!

**Comment by gws** [comment]
OK, I understand completely!

If we do design new APIs that return 4xx errors to clients, I think the body should contain enough information for the client to be able to act on it and potentially retry the request if desired (just like a human would do with a failed form validation), and not necessarily be forced to give up forever or complain loudly (the first time).

:+1: :ship:

**Comment by cap10morgan** [comment]
I ended up obviating the need for this.

**Pull request closed without merging**
